Add `constraints` format-clause with support for UNIQUE and PRIMARY KEY

drew-patel opened the pull request:
Related issue - #15

- new constraints `defhelper` which adds a top-level `:constraints` key to the complete-sql-map
- moved format-clause for `:with-columns` into its own function so that it can be reused in the `constraints` format-clause method as well
- when constraints aren't present, the formatter uses `:with-columns` to construct the SQL string
- when constraints are present, the formatter uses `:constraints` to construct the columns and constraints string wrapped in the same set of parens

drew-patel committed:
Create `constraints` format-clause with primary-key + unique support

Requires `:with-columns` to have its format clause pulled out
into its own function so that it can be used with `constraints`.

When `constraints` are present, the columns are constructed by
the `constraints` formatter so that columns and constraints can
be wrapped by the same set of parens.

drew-patel committed:
Add tests which create a table with primary-key & unique constraints

scimetfoo commented on src/honeysql_postgres/helpers.cljc:
```diff
@@ -38,6 +39,17 @@
 (defhelper with-columns [m args]
   (assoc m :with-columns args))
 
+(defhelper constraint [m args]
```
A couple of suggestions here

1. Use `defhelper` only to update the query map with the constraint clauses and not to construct the SQL function call. We want to utilise the format-clause multimethod to construct the function call so that it works with [honeysql's priority-based ordering rule system](https://github.com/seancorfield/honeysql/blob/develop/src/honeysql/format.cljc#L262). Also, add a [priority](https://github.com/nilenso/honeysql-postgres/blob/master/src/honeysql_postgres/format.cljc#L7) to the constraint clause, which I think would be a number slightly larger than the priority assigned for `with-columns`. I just noticed that `with-columns` doesn't have a priority assigned to it either,  but it will need one now. Could you please add one for it as well? 
2. Pluralise this function's name and have it support multiple constraints so that going forward we're able to use this function for defining multiple constraints.  ```
(constraints [:unique [:product_no :product_sku]]
                    [:primary-keys [:name]])
```
3. Make constraints a top-level key in the query map instead of `assoc`ing it with the columns. It otherwise ends up treating the constraint as a column.

scimetfoo commented:
@drew-patel thanks for submitting this PR!

A couple of suggestions:

1. Use `defhelper` only to update the query map with the constraint clauses and not to construct the SQL function call. We want to utilise the `format-clause` multimethod to construct the function call so that it works with [honeysql's priority-based ordering rule system](https://github.com/seancorfield/honeysql/blob/develop/src/honeysql/format.cljc#L262). Also, add a [priority](https://github.com/nilenso/honeysql-postgres/blob/master/src/honeysql_postgres/format.cljc#L7) to the constraint clause, which I think would be a number slightly larger than the priority assigned to `with-columns`. I just noticed that `with-columns` doesn't have a priority assigned to it either,  but it will need one now. Could you please add one for it as well? 
2. Pluralise `constraint` so that going forward we're able to use this function for defining multiple constraints. Eg:
```
(constraints [:unique [:product_no :product_sku]]
             [:primary-key [:name]])
```
3. Add a top-level key called `:constraints` to the query map instead of `assoc`ing the constraints to the collection of columns. It otherwise ends up treating each constraint as a column.
4. Support multiple constraints? It could also just be multiple unique constraints to being with.

drew-patel commented:
@scimetfoo Thanks for the feedback and suggestions! 

All the points are clear. I'll ask questions if I'm unsure about anything while working on them. Will get back to you in a few days with the changes.

drew-patel commented:
@scimetfoo I have addressed comments 1-3 as described above. I've only implemented the solution for `UNIQUE` and `PRIMARY KEY` so far since they behave similarly. 

- I have no idea why I'm getting an extra space after the table name. See [one of the tests](https://github.com/nilenso/honeysql-postgres/pull/60/files#diff-568a78dc249a6592ba8f8e38bdfe6a2c3c06531fecdc7e549b5d060368c9ee56R151).

- Also, I'm not sure if the priority numbers chosen are ok. Happy to fix those if they aren't.

Thanks!

scimetfoo commented on src/honeysql_postgres/format.cljc:
```diff
@@ -143,11 +145,40 @@
+
+(defmethod format-clause :constraints [[_ constraints-seq] complete-sql-map]
+  (let [columns (:with-columns complete-sql-map)]
+    (when (and (seq columns)
```
Filtering constraints before formatting the columns and constraints would form the basis for adding a trailing comma after the columns and also between constraints. If the validation of the constraint type is done after adding the trailing comma, and an unsupported constraint is found, the formatted query produced is invalid. Eg:

```
(-> (create-table :products)
    (with-columns [[:product_no :integer]
                   [:name :text]
                   [:product_sku :integer]])
    (constraints [[:foreign-key [:product_no]]])
    sql/format)
;; => ["CREATE TABLE products  (product_no integer, name text, product_sku integer, )"]
```
This also happens if there is more than one constraint for the same reason as this:

```
(clojure.string/join ", " ["a" nil "b"])
;; => "a, , b"
```

The nil here is added in when [this](https://github.com/nilenso/honeysql-postgres/pull/60/files#diff-f81b5b6904b08f9c50a240ecab26f660e4f13c113064b4519324a08f69ac62f8R155) isn't satisfied.

I propose filtering out all the invalid / unsupported constraints before formatting them. 

```
(defmethod format-clause :constraints [[_ [constraints]] complete-sql-map]
  (let [[columns]   (:with-columns complete-sql-map)
        constraints (filter (fn [[constraint-type constraint-args]]
                              (contains? constraints-format-map constraint-type)) constraints)]
    (sqlf/paren-wrap
     (str (when (seq columns) (format-columns-clause columns))
          (when (seq constraints) (str ", " 
                                                        (sqlf/comma-join (map format-constraint-clause
                                                                                              constraints))))))))
```

drew-patel replied:
Ahh! Thanks for that. I should have paid attention to trailing commas. I have done what you have suggested in the code block above and it works.

I've added a 4th test case as well which ignores invalid/unsupported constraints.

scimetfoo commented on src/honeysql_postgres/format.cljc:
```diff
@@ -143,11 +145,40 @@
+
+(defn- format-constraint-clause
+  [constraint]
+  (let [[constraint-type constraint-args] constraint]
```
It would be nicer if the sequential destructuring was done in the params.

drew-patel replied:
Done!

scimetfoo reviewed (commented):
Looks good. Please add examples to the readme.

drew-patel committed:
Update README to include constraints examples (unique & primary-key)

drew-patel commented:
@scimetfoo Please let me know if the examples are sufficient. Thanks again for the review!
